Updated the database privilege syntax to show correct optional parts

Hunterness opened the pull request:
Also removed untrue statement about multiple databases being allowed and added link to two image sources (images changed in manual modeling https://github.com/neo-technology/neo4j-manual-modeling/pull/977 and https://github.com/neo-technology/neo4j-manual-modeling/pull/983)

Hunterness committed:
Updated the database privilege syntax to show correct optional parts

and removed untrue statement about multiple databases being allowed

Also added link to image source (image changed in manual modeling)

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/database/admin-database-syntax.asciidoc:
```diff
@@ -5,74 +5,74 @@
 GRANT ACCESS
-    ON DATABASE name
+    ON DATABASE[S] {name \| *}
     TO role
```
Maybe add `[, role2, ...]` to all the role parts to show they can be multiple

Lojjs replied:
I don't think it can be multiple roles in this place right? I thought that was only for `GRANT ROLE r1, r2 TO USER u1, u2, u3`. Or have we added this functionality recently?

Hunterness replied:
I don't think it is recent, but I'm pretty sure you can...

One of the parser tests: 
```
test(s"$command $privilege ON DATABASE * $preposition role1, role2") {
     yields(privilegeFunc(action, ast.AllGraphsScope() _, Seq("role1", "role2")))
}
```

Hunterness commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityAdministrationTest.scala:
```diff
@@ -58,6 +58,7 @@ class SecurityAdministrationTest extends DocumentingTest with QueryStatisticsTes
+      // image source: https://docs.google.com/drawings/d/1Zc5eawW58r4CI8e4FeYnA8cwUyIJX-7B4Eecj-qkd_8/edit
       section("The database `ACCESS` privilege", "administration-security-administration-database-access", "enterprise-edition") {
         p(
           """The `ACCESS` privilege can be used to enable the ability to access a database.
```
Further down in this file (under `Granting all database administration privileges`) the example text lists the ability to grant most but not all privileges that will be granted, maybe it should?

Lojjs replied:
Yes, I think that is a good idea.

Hunterness committed:
Fixed review comments

Added multiple roles to the commands
More explicit info about granting all database privilege
Fixed missed table
Removed DENY ACCESS from table since no other commands had DENY there, only grant
Removed todo and out-commented line

Hunterness committed:
Updated syntax for graph privileges and role management as well

Adding multiple roles and users
Adding `*` alternative to graph name

Hunterness committed:
Added syntax table for the role management privileges as well

Since the chapter only had examples before

Hunterness committed:
Added source for updated graph privilege image as well

Lojjs commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityAdministrationTest.scala:
```diff
@@ -196,7 +196,7 @@ class SecurityAdministrationTest extends DocumentingTest with QueryStatisticsTes
 
         p(
-          """For example, granting the ability to create indexes, constraints, labels, relationship types and property names on the database `neo4j` to the role `regularUsers` is done using the following query.""".stripMargin)
+          """For example, granting the ability to access, start and stop all databases and create indexes, constraints, labels, relationship types and property names on the database `neo4j` to the role `regularUsers` is done using the following query.""".stripMargin)
```
If you are listing all privileges here I do not think it should say `For example`?

Hunterness replied:
I think the `For example` is because after is an example query? (specifically giving it on `neo4j` to `regularUsers`)

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/grant-deny-syntax.asciidoc:
```diff
@@ -9,24 +9,24 @@ The syntax is:
 | [source, cypher]
-GRANT graph-privilege ON GRAPH dbname entity TO role
-| Grant a privilege to a role
+GRANT graph-privilege ON GRAPH[S] {name \| *} entity TO roles
```
entity is optional as well

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/grant-deny-syntax.asciidoc:
```diff
@@ -67,8 +67,8 @@ this privilege can only be assigned to all nodes, relationships and properties i
 *** `ELEMENTS` label (both nodes and relationships).
 ** The label or type can be `+*+` which means all labels or types.
 ** Multiple labels or types can be specified, comma-separated.
```
Add an item about default value if not specified

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/grant-deny-syntax.asciidoc:
```diff
@@ -52,7 +52,7 @@ this combines both TRAVERSE and READ allowing an entity to be found and its prop
 +
 ** `WRITE +{*}+`
```
No longer true, should not have `{*}`

Hunterness committed:
Updated the entity part of the graph privileges to show as optional

and added a list item about the default value if omitted.

Also removed `{*}` from WRITE since it is no longer true

AlexicaWright commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/database/admin-role-database.asciidoc:
```diff
@@ -9,24 +9,24 @@ The syntax is:
-| Grant a privilege to a role
+GRANT database-privilege ON DATABASE[S] {dbname \| *} TO roles
+| Grant a privilege to one or multiple roles
 
```
**roles**, does that imply that it is always more than one, or can it denote a single role as well? This comment applies to the following examples as well.
A suggestion to avoid ambiguity would be to use this format: `role[, ...]` as used above and below.

Hunterness replied:
The reason why I chose `roles` here was to match the syntax for the graph privileges, where they have this annotation for other parts of the commands (at least properties/labels in the individual commands). I had the `role[, ...]` syntax here before updating the graph privilege commands, and can change it back here.

AlexicaWright commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/database/admin-role-database.asciidoc:
```diff
@@ -93,12 +93,11 @@ allows all of the above privileges to be enabled for the specified database.
   Any new databases created after this command will also be associated with these privileges.
-* _role_
-** The role or roles to associate the privilege with.
+* _roles_
```
see above comment about **roles**

AlexicaWright reviewed (commented):
Overall it looks very good to me. However, there is the question about **role/roles**. I don't know if it is possible and/or applicable to use _role[...]_ instead of _roles_?

Hunterness committed:
Update the role/roles syntax for database privileges

Kept it for graph privileges to be consistent in that section
